codejail/jail.py

```diff
@@ -0,0 +1,329 @@
+            tmptmp = os.path.join(homedir, "tmp")
+            os.mkdir(tmptmp)
+            os.chmod(tmptmp, 0777)
+
+            argv = argv or []
+            env = {'TMPDIR': 'tmp'}
+
+            # All the supporting files are copied into our directory.
+            for filename in files or ():
+                dest = os.path.join(homedir, os.path.basename(filename))
+                if os.path.islink(filename):
+                    os.symlink(os.readlink(filename), dest)
+                elif os.path.isfile(filename):
+                    shutil.copy(filename, homedir)
+                else:
+                    shutil.copytree(filename, dest, symlinks=True)
+
+            # Create the main file.
+            if code:
+                with open(os.path.join(homedir, "jailed_code"), "wb") as jailed:
+                    jailed.write(code)
+
+                argv = ["jailed_code"] + argv
+
+            # Create extra files requested by the caller:
+            for name, content in extra_files or ():
+                with open(os.path.join(homedir, name), "wb") as extra:
+                    extra.write(content)
+
+            cmd = []
+
+            # Build the command to run.
+            if self.user:
+                # Run as the specified user
+                cmd.extend(['sudo', '-u', self.user, 'TMPDIR=tmp'])
+
+            # Add the command-specific pieces
+            cmd.extend(self.cmdline_start)
+            # Add the code-specific command line pieces.
+            cmd.extend(argv)
+
+            # Use the configuration and maybe an environment variable to determine
+            # whether to use a proxy process.
+            use_proxy = limits.LIMITS["PROXY"]
+            if use_proxy is None:
+                use_proxy = int(os.environ.get("CODEJAIL_PROXY", "0"))
+            if use_proxy:
+                run_subprocess_fn = run_subprocess_through_proxy
+            else:
+                run_subprocess_fn = run_subprocess
+
+            # Run the subprocess.
+            status, stdout, stderr = run_subprocess_fn(
+                cmd=cmd, cwd=homedir, env=env, slug=slug,
+                stdin=stdin,
+                realtime=limits.LIMITS["REALTIME"], rlimits=limits.create_rlimits(),
+            )
+
+            # TODO: limit too much stdout/stderr data?
+            result = JailResult(status, stdout, stderr)
+
+            # Remove the tmptmp directory as the sandbox user since the sandbox
+            # user may have written files that the application user can't
+            # delete.
+            run_subprocess_fn(self._build_rm_command(tmptmp), cwd=homedir)
+
+        return result
+
+    def _build_rm_command(self, directory):
+        """
+        Collect the command line arguments to remove the specified directory
+        """
+        rm_cmd = []
+        if self.user:
+            rm_cmd.extend(['sudo', '-u', self.user])
+        rm_cmd.extend([
+            '/usr/bin/find', directory,
```

Comment:
I think there might be issues with order of deletion if you don't use -r (trying to delete directories before files), and with trying to access already deleted code if you use -r but don't limit the depth.  Either way, this was inherited code, and I think changing it is outside the scope of this ticket.